Add admin API to send trace notifications to registered

poornas opened the pull request:
listeners

<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
Currently all trace is captured via setting MINIO_HTTP_TRACE to output file . Adding Admin API for registering /removing trace listeners for sending trace to all registered listeners. This trace will be displayed on mc with ``` mc admin trace myminio ``` 

Leaving tracing to output file in place for the time being until this functionality is robust
 
## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## Regression
<!-- Is this PR fixing a regression? (Yes / No) --> No
<!-- If Yes, optionally please include minio version or commit id or PR# that caused this regression, if you have these details. -->

## How Has This Been Tested? 
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
manually in conjunction with mc admin trace command 
## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

harshavardhana commented on cmd/peer-rpc-client-target.go:
```diff
@@ -1,5 +1,5 @@
 /*
- * Minio Cloud Storage, (C) 2018 Minio, Inc.
+ * Minio Cloud Storage, (C) 2019 Minio, Inc.
```
2018 and 2019

harshavardhana commented on cmd/peer-rpc-client-trace-target.go:
```diff
@@ -0,0 +1,52 @@
+/*
+ * Minio Cloud Storage, (C) 2018 Minio, Inc.
```
This should be 2019

harshavardhana commented on pkg/trace/trace.go:
```diff
@@ -0,0 +1,220 @@
+	const timeFormat = "2006-01-02 15:04:05 -0700"
+	ri := ti.ReqInfo
+	rs := ti.RespInfo
+	msg := fmt.Sprintf("[REQUEST %s] [%s] [%s]\n", ti.Name, ti.ReqID, ri.Time.Format(timeFormat))
```
Rather concatenate let's either use strings.WriteString etc.

donatello commented on cmd/peer-rpc-client-target.go:
```diff
@@ -1,5 +1,5 @@
 /*
- * Minio Cloud Storage, (C) 2018 Minio, Inc.
+ * Minio Cloud Storage, (C) 2018, 2019 Minio, Inc.
```
Not needed as there are no other changes.

harshavardhana commented on cmd/peer-rpc-server.go:
```diff
@@ -397,5 +398,66 @@ func registerPeerRPCRouter(router *mux.Router) {
+
+// ListenTraceNotification - handles listen bucket notification RPC call.
+// It creates PeerRPCClient target which pushes requested events to target in remote peer.
+func (receiver *peerRPCReceiver) ListenTraceNotification(args *ListenTraceNotificationArgs, reply *VoidReply) error {
```
One problem of doing rpc for trace over the network might be very chatty, not sure if it might impact the server throughput. We may need to say if trace is enabled it may perhaps reduce the overall cluster performance.

poornas replied:
that's true

harshavardhana replied:
We should discuss this..

donatello commented on cmd/globals.go:
```diff
@@ -33,7 +33,7 @@ import (
 	"github.com/minio/minio/pkg/certs"
 	"github.com/minio/minio/pkg/dns"
-	"github.com/minio/minio/pkg/iam/policy"
+	iampolicy "github.com/minio/minio/pkg/iam/policy"
```
I dont understand how this can change in the file and then no other lines are changed to use this module renaming.

donatello reviewed (commented):
LGTM - its a straightforward implementation. I'd like some steps to test with mc so I can run and try this PR.

poornas committed:
Add admin API to send trace notifications
to registered listeners.

Remove current functionality to log trace to file
using MINIO_HTTP_TRACE env, and replace it with
mc admin trace command on mc client.

donatello commented on pkg/pubsub/pubsub.go:
```diff
@@ -0,0 +1,73 @@
+	for i, sub := range ps.subs {
+		if sub == ch {
+			ps.subs = append(ps.subs[:i], ps.subs[i+1:]...)
+		}
```
Channel being removed can also be closed here?

donatello commented on cmd/httptrace.go:
```diff
@@ -0,0 +1,183 @@
+		ch := ps.Subscribe()
+		go func(client *peerRESTClient, ch chan interface{}) {
+			for {
+				msg := <-ch
```
Why not `for msg := range ch {` here?

donatello commented on cmd/httptrace.go:
```diff
@@ -0,0 +1,183 @@
+	ch := ps.Subscribe()
+	go func(pss *HTTPTraceSys, ch chan interface{}) {
+		for {
+			msg := <-ch
```
Same q - range over channel?

donatello commented on cmd/peer-rest-server.go:
```diff
@@ -547,6 +549,87 @@ func (s *peerRESTServer) SignalServiceHandler(w http.ResponseWriter, r *http.Req
+	}
+
+	var args traceListenerReq
+	if r.ContentLength <= 0 {
```
Is this check needed? Won't it usually just be correct?

krishnasrinivas commented on pkg/handlers/http-tracer.go:
```diff
@@ -109,92 +113,73 @@ func (r *recordResponseWriter) Flush() {
 		}
+	} else {
+		rq.Body = fmt.Sprintln(string(bodyPlaceHolder))
+	}
```
above if else can be simplified by
```
+       if logBody {
+               rq.Body = reqBodyRecorder.Data()
+       } else {
+               rq.Body = bodyPlaceHolder
+       }
```

krishnasrinivas commented on pkg/handlers/http-tracer.go:
```diff
@@ -109,92 +113,73 @@ func (r *recordResponseWriter) Flush() {
+	var reqBodyRecorder *recordRequest
+
+	// Generate short random request ID
+	reqID := fmt.Sprintf("%f", float64(time.Now().UnixNano())/1e10)
```
reqID can be removed as we always print request/response in one shot (i.e there is no intermingling of requests and responses)

krishnasrinivas commented on pkg/handlers/http-tracer.go:
```diff
@@ -109,92 +113,73 @@ func (r *recordResponseWriter) Flush() {
+	reqBodyRecorder = &recordRequest{Reader: r.Body, logBody: logBody}
+	r.Body = ioutil.NopCloser(reqBodyRecorder)
+
+	if logNodeName {
```
log node names always (since this info is for `mc admin` command and not for subnet)

krishnasrinivas commented on pkg/trace/trace.go:
```diff
@@ -0,0 +1,220 @@
+	Headers    map[string]string
+	Body       string
+	StatusCode int
+	ErrMsg     string
```
this field is not needed

krishnasrinivas commented on pkg/trace/trace.go:
```diff
@@ -0,0 +1,220 @@
+	Host    string
+	URL     url.URL
+	Headers map[string]string
+	Body    string
```
instead:
```
+       Time     time.Time
+       Method   string
+       Path     string
+       RawQuery string
+       Headers  map[string]string
+       Body     []byte
```

Body is []byte`

krishnasrinivas replied:
need to put r.Host r.TransferEncoding r.ContentLength as part of `Headers`

krishnasrinivas commented on pkg/trace/trace.go:
```diff
@@ -0,0 +1,220 @@
+	"sync"
+	"time"
+)
+
```
all the fields below needs json tags

poornas replied:
fixed

krishnasrinivas commented on pkg/handlers/http-tracer.go:
```diff
@@ -109,92 +113,73 @@ func (r *recordResponseWriter) Flush() {
+			// If there was no error response and body logging is disabled
+			// then we print <BODY> as a placeholder for the actual body.
+			rs.Body = fmt.Sprintln(string(bodyPlaceHolder))
+		}
```
above if else can be simplified as:
```
+       if logBody {
+               rs.Body = respBodyRecorder.Body()
+       } else {
+               rs.Body = bodyPlaceHolder
+       }
```

krishnasrinivas commented on cmd/handler-utils.go:
```diff
@@ -326,18 +326,28 @@ func extractPostPolicyFormValues(ctx context.Context, form *multipart.Form) (fil
+			return
+		}
+		logNodeName := globalIsDistXL && !globalCLIContext.Anonymous
+		trace, _ := httptracer.Trace(f, false, logNodeName, w, r)
```
looks like httptracer.Trace() always returns `nil` for error in which case the function need not return error value

krishnasrinivas commented on cmd/handler-utils.go:
```diff
@@ -326,18 +326,28 @@ func extractPostPolicyFormValues(ctx context.Context, form *multipart.Form) (fil
-	if globalHTTPTraceFile == nil {
-		return f
+	return func(w http.ResponseWriter, r *http.Request) {
+		if globalTrace == nil || !globalTrace.HasTraceListeners() {
```
globalTrace should always be initialized hence no need to check for `nil` here

krishnasrinivas commented on cmd/httptrace.go:
```diff
@@ -0,0 +1,114 @@
+}
+
+// Trace writes http trace to writer
+func (sys *HTTPTraceSys) Trace(targetID string, traceCh chan trace.Info, doneCh chan struct{}, trcAll bool) {
```
Usual pattern is this function should return `traceCh chan trace.Info`

krishnasrinivas commented on cmd/admin-handlers.go:
```diff
@@ -1415,3 +1416,40 @@ func (a adminAPIHandlers) SetConfigKeysHandler(w http.ResponseWriter, r *http.Re
+
+	globalTrace.Trace(targetID, traceCh, doneCh, trcAll)
+	defer globalTrace.Unsubscribe(targetID)
+	defer globalTrace.UnsubscribePeers(targetID)
```
Unsubscribe, UnsubscribePeers are not needed. It can be done in a simpler way (explained in peer-rest-client.go)

krishnasrinivas commented on cmd/peer-rest-client.go:
```diff
@@ -359,6 +363,64 @@ func (client *peerRESTClient) SignalService(sig serviceSignal) error {
 	return nil
 }
 
+// Trace - send http trace request to peer nodes
```
you can avoid doing out of band unsubscribe like this:
```
+type traceEndReader struct {
+       doneCh chan struct{}
+}
+
+func (t *traceEndReader) Read(p []byte) (int, error) {
+       <-t.doneCh
+       return 0, io.EOF
+}
+
 // Trace - send http trace request to peer nodes
 func (client *peerRESTClient) Trace(targetID string, doneCh chan struct{}, trcAll bool) (chan trace.Info, error) {
        ch := make(chan trace.Info)
@@ -374,7 +383,7 @@ func (client *peerRESTClient) Trace(targetID string, doneCh chan struct{}, trcAl
                        values.Set(peerRESTTraceAll, strconv.FormatBool(trcAll))
                        values.Set(peerRESTTraceTargetID, targetID)
 
-                       respBody, err := client.call(peerRESTMethodTrace, values, &reader, -1)
+                       respBody, err := client.call(peerRESTMethodTrace, values, &traceEndReader{doneCh}, -1)
 
                        if err != nil {
                                //retry
@@ -383,8 +392,7 @@ func (client *peerRESTClient) Trace(targetID string, doneCh chan struct{}, trcAl
 
                        bio := bufio.NewScanner(respBody)
 
-                       // Close the response body.
-                       defer respBody.Close()
+                       defer http.DrainBody(respBody)
 
                        // Unmarshal each line, returns marshaled values.
                        for bio.Scan() {
```

krishnasrinivas commented on cmd/peer-rest-server.go:
```diff
@@ -547,6 +549,38 @@ func (s *peerRESTServer) SignalServiceHandler(w http.ResponseWriter, r *http.Req
+		}
+		w.(http.Flusher).Flush()
+	}
+}
```
You can avoid out of band Unsubscribe rest call by doing this
```
func (s *peerRESTServer) TraceHandler(w http.ResponseWriter, r *http.Request) {
        trcAll := false
        if a := r.URL.Query().Get("all"); a != "false" {
                trcAll = true
        }
        doneCh := make(chan struct{})
        go func() {
                io.Copy(ioutil.Discard, r.Body)
                close(doneCh)
        }()
        ch := globalTrace.pubsub.Subscribe()
        defer globalTrace.pubsub.Unsubscribe(ch)

        for {
                select {
                case <-doneCh:
                        return
                case entry := <-ch:
                        trcInfo := entry.(trace.Info)
                        // omit inter-node traffic if trcAll is false
                        if !trcAll && strings.HasPrefix(trcInfo.ReqInfo.URL.Path, "/minio") {
                                continue
                        }
                        data, err := json.Marshal(trcInfo)
                        if err != nil {
                                return
                        }
                        data = append(data, byte('\n'))
                        if _, err = w.Write(data); err != nil {
                                return
                        }
                        w.(http.Flusher).Flush()
                }
        }
}
```

krishnasrinivas replied:
Also in the above for loop, you can select on `GlobalServiceDoneCh` and remove the function stopHTTPTrace

krishnasrinivas commented on cmd/httptrace.go:
```diff
@@ -0,0 +1,114 @@
+type HTTPTraceSys struct {
+	peers        []*peerRESTClient
+	pubsub       *pubsub.PubSub
+	traceTargets map[string]chan interface{}
```
`traceTargets` will no longer be needed (because of the way unsubscribe is done as explained in peer-rest-client.go)

krishnasrinivas commented on pkg/pubsub/pubsub.go:
```diff
@@ -0,0 +1,106 @@
+}
+
+// UnsubscribeAll removes all subscribers
+func (ps *PubSub) UnsubscribeAll() {
```
This is not needed anymore as each subscriber calls Unsubscribe()

krishnasrinivas commented on pkg/pubsub/pubsub.go:
```diff
@@ -0,0 +1,106 @@
+}
+
+// HasSubscribers returns true if pubsub system has subscribers
+func (ps *PubSub) HasSubscribers() bool {
```
needs locking

krishnasrinivas commented on pkg/pubsub/pubsub.go:
```diff
@@ -0,0 +1,106 @@
+	if count <= bufferCount {
+		idx := count % bufferCount
+		ps.buf[idx] = item
+	}
```
buffered chans should be used instead of using custom `[] ps.buf`
Also a go routine should be used to send to the buffered channel so that `Publish()` returns immediately

krishnasrinivas replied:
you can create buffered chans for publisher chan and subcriber chan

krishnasrinivas commented on pkg/trace/trace.go:
```diff
@@ -0,0 +1,68 @@
+	Path     string    `json:"path,omitempty"`
+	RawQuery string    `json:"rawquery,omitempty"`
+	// Host     string
+	URL     url.URL     `json:"url,omitempty"`
```
`URL` can be removed

krishnasrinivas commented on pkg/trace/trace.go:
```diff
@@ -0,0 +1,68 @@
+	"time"
+)
+
+// Short trace record
```
These structs should be part of  pkg/madmin so that client `mc` need not import server related code which is present in `pkg/trace`

Commit:
fix review comments

krishnasrinivas commented on cmd/peer-rest-server.go:
```diff
@@ -547,6 +549,39 @@ func (s *peerRESTServer) SignalServiceHandler(w http.ResponseWriter, r *http.Req
+	trcAll := false
+	if a := r.URL.Query().Get(peerRESTTraceAll); a != "false" {
+		trcAll = true
+	}
```
trcAll := r.URL.Query().Get(peerRESTTraceAll) == "true"

harshavardhana commented on cmd/httptrace.go:
```diff
@@ -0,0 +1,110 @@
+/*
+ * Minio Cloud Storage, (C) 2019 Minio, Inc.
```
```suggestion
 * MinIO Cloud Storage, (C) 2019 Minio, Inc.
```

harshavardhana commented on cmd/peer-rest-server.go:
```diff
@@ -547,6 +549,46 @@ func (s *peerRESTServer) SignalServiceHandler(w http.ResponseWriter, r *http.Req
+		case entry := <-ch:
+			trcInfo := entry.(trace.Info)
+			// omit inter-node traffic if trcAll is false
+			if !trcAll && strings.HasPrefix(trcInfo.ReqInfo.Path, "/minio") {
```
```suggestion
			if !trcAll && strings.HasPrefix(trcInfo.ReqInfo.Path, minioReservedBucketPath) {
```

harshavardhana commented on cmd/peer-rest-server.go:
```diff
@@ -588,5 +630,7 @@ func registerPeerRESTHandlers(router *mux.Router) {
 
 	subrouter.Methods(http.MethodPost).Path("/" + peerRESTMethodReloadFormat).HandlerFunc(httpTraceHdrs(server.ReloadFormatHandler)).Queries(restQueries(peerRESTDryRun)...)
 
+	subrouter.Methods(http.MethodPost).Path("/" + peerRESTMethodTrace).HandlerFunc(server.TraceHandler)
```
Adding new handlers, please bump the version

harshavardhana commented on pkg/madmin/README.md:
```diff
@@ -537,3 +537,22 @@ __Example__
+
+<a name="ListenTrace"></a>
+### ListenTrace(allTrace bool,doneCh <-chan struct{}) <-chan TraceInfo
+get http trace from minio server
```
We should explain a bit more here, that how it is that trace comes from all servers in a cluster etc. Start the sentence with a capitalization since this is a document.

harshavardhana commented on pkg/madmin/api-trace.go:
```diff
@@ -0,0 +1,93 @@
+/*
+ * Minio Go Library for Amazon S3 Compatible Cloud Storage
```
This is an incorrect license, it's not `Minio` Go library. Also, Minio should be MinIO

harshavardhana commented on pkg/madmin/api-trace.go:
```diff
@@ -0,0 +1,93 @@
+
+// TraceInfo holds http trace
+type TraceInfo struct {
+	Trc trc.Info
```
any reason to remove "a" from Trc, it is better to keep it Trace I believe, unless without a strong argument.

harshavardhana commented on pkg/madmin/examples/trace.go:
```diff
@@ -0,0 +1,52 @@
+// +build ignore
+
+/*
+ * MinIO Cloud Storage, (C) 2017 MinIO, Inc.
```
The license should be 2019 since this a new file.

harshavardhana commented on pkg/pubsub/pubsub.go:
```diff
@@ -0,0 +1,82 @@
+/*
+ * Minio Cloud Storage, (C) 2019 Minio, Inc.
```
```suggestion
 * MinIO Cloud Storage, (C) 2019 Minio, Inc.
 
```

harshavardhana commented on pkg/trace/trace.go:
```diff
@@ -0,0 +1,54 @@
+/*
+ * Minio Cloud Storage, (C) 2019 Minio, Inc.
```
```suggestion
 * MinIO Cloud Storage, (C) 2019 Minio, Inc.
```

harshavardhana replied:
```suggestion
 * MinIO Cloud Storage, (C) 2019 Minio, Inc.
```

harshavardhana commented on cmd/admin-handlers.go:
```diff
@@ -65,7 +65,7 @@ var (
 	// the server uses.
 	adminAPIVersionInfo = madmin.AdminAPIVersionInfo{
-		Version: "1",
+		Version: "2",
```
Actually, this API is not used, let's not change it keep it as "1" - actual change should be in admin/v1 to admin/v2 in the router. It's not needed since its just new API.

harshavardhana commented on pkg/pubsub/pubsub.go:
```diff
@@ -0,0 +1,82 @@
+	"sync"
+)
+
+//PubSub holds publishers and subscribers
```
Space after `//` 
```suggestion
// PubSub holds publishers and subscribers
```

harshavardhana commented on pkg/pubsub/pubsub.go:
```diff
@@ -0,0 +1,82 @@
+/*
+ * MinIO Cloud Storage, (C) 2019 MinIO, Inc.
```
Space after headling
```suggestion
 * MinIO Cloud Storage, (C) 2019 MinIO, Inc.
 
```

harshavardhana commented on pkg/pubsub/pubsub_test.go:
```diff
@@ -0,0 +1,65 @@
+/*
+ * MinIO Cloud Storage, (C) 2019 MinIO, Inc.
```
Add a space after the headline
```suggestion
 * MinIO Cloud Storage, (C) 2019 MinIO, Inc.
 
```

harshavardhana commented on pkg/madmin/README.md:
```diff
@@ -537,3 +537,22 @@ __Example__
+
+<a name="Trace"></a>
+### Trace(allTrace bool,doneCh <-chan struct{}) <-chan TraceInfo
+Ask all nodes in minio cluster to send trace of HTTP requests
```
```suggestion
Enable HTTP request tracing on all nodes in a MinIO cluster
```

harshavardhana committed:
Merge branch 'master' into tr
